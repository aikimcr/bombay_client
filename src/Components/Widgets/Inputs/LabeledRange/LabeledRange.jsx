import React, { useImperativeHandle, useRef } from 'react';
import PropTypes from 'prop-types';
import { RangeInput } from '.';

import './LabeledRange.scss';

export const LabeledRange = ({ labelText, inputId, inputProps }) => {
  return (
    <div className="labeled-range" data-testid="labeled-range">
      <label htmlFor={inputId}>{labelText}</label>
      <RangeInput {...inputProps} id={inputId} />
    </div>
  );
};

LabeledRange.propTypes = {
  fieldName: PropTypes.string.isRequired,
  labelText: PropTypes.string.isRequired,
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
  onChange: PropTypes.func,
};

export default LabeledRange;
