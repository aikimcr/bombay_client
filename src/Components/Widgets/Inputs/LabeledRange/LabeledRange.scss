.labeled-range {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  gap: 1rem;
  width: 100%;

  & > .labeled-range__label {
    min-width: max-content;
    max-width: 12ch;
    width: 100%;
  }
}