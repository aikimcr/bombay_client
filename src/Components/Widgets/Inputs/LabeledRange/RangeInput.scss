.range-input {
  display: grid;
  grid-auto-flow: column;
  grid-template-columns: 1fr auto;
  align-items: center;
  justify-content: start;
  gap: .5rem;

  &__inputs {
    display: grid;
    grid-auto-flow: row;
  }

  &__input {
    padding: .25rem 1rem;
    border-radius: 1rem;
  }
}