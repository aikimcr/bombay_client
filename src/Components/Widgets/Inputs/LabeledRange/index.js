export * from './LabeledRange';
export * from './RangeInput';
