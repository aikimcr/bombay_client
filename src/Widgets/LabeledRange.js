import PropTypes from 'prop-types'
import { useRef } from 'react';
import CloseOrClear from './CloseOrClear';

import './LabeledInput.scss';

function LabeledRange(props) {
    const topRef = useRef(null);

    const modelId = props.model ? props.model.get('id') : 'isNew';
    const inputId = `${props.modelName}-${props.fieldName}-${modelId}`;
    const defaultValue = props.model ? props.model.get(props.fieldName) : '';

    function changeHandler(evt) {
        evt.preventDefault();

        const newValue = evt.target.value;

        const fields = topRef.current.querySelectorAll('input');
        fields.forEach(field => { field.value = newValue });
    }

    function clearValue(evt) {
        evt.preventDefault();

        const fields = topRef.current.querySelectorAll('input');
        fields.forEach(field => { field.value = null });
    }

    return (
        <div ref={topRef} className='labeled-range' data-modelname={props.modelName} data-targetfield={props.fieldName}>
            <label htmlFor={inputId}>{props.labelText}</label>
            <div className='range-input'>
                <input id={inputId} type='range' min={props.min} max={props.max} defaultValue={defaultValue} onChange={changeHandler}></input>
                <input name={props.fieldName} type='number' defaultValue={defaultValue} min={props.min} max={props.max} onChange={changeHandler}></input>
                <CloseOrClear onClick={clearValue} />
            </div>
        </div>
    )
}

LabeledRange.propTypes = {
    modelName: PropTypes.string.isRequired,
    fieldName: PropTypes.string.isRequired,
    labelText: PropTypes.string.isRequired,
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    model: PropTypes.object,
    onChange: PropTypes.func,
}

export default LabeledRange