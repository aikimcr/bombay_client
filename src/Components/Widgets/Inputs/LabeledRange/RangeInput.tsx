import React, { useRef } from 'react';
import { debounce } from 'lodash';

import './RangeInput.scss';
import { isHTMLInputElement } from '../../../../Utilities';
import { ClearButton } from '../../Buttons';

export interface RangeInputProps {
  defaultValue: number;
  id: string;
  max: number;
  min: number;
  name: string;
  onChange: (newValue: number) => void;
  ref: ReturnType<typeof useRef<HTMLInputElement>>;
}

export const RangeInput: React.FC<RangeInputProps> = ({
  defaultValue,
  id,
  max,
  min,
  name,
  onChange,
  ref,
}): React.ReactElement => {
  const sliderRef = useRef<HTMLInputElement>(null);
  const inputRef = ref || useRef<HTMLInputElement>(null);

  const setValues = (newValue: string | number) => {
    const newValueString = newValue ? newValue.toString() : null;

    if (sliderRef.current) {
      sliderRef.current.value = newValueString;
    }

    if (inputRef.current) {
      inputRef.current.value = newValueString;
    }

    onChange?.call(null, parseInt(newValueString));
  };

  const changeHandler: React.ChangeEventHandler<HTMLInputElement> = (evt) => {
    evt.preventDefault();

    if (isHTMLInputElement(evt.target)) {
      setValues(evt.target.value);
    }

    return evt.target;
  };

  const debouncedChangeHandler: React.ChangeEventHandler<HTMLInputElement> =
    debounce(changeHandler, 500);

  const clearValue = () => {
    setValues(defaultValue);
  };

  return (
    <div className="range-input" data-testid="range-input">
      <div className="range-input__inputs">
        <input
          id={id}
          className="range-input__slider"
          ref={sliderRef}
          type="range"
          min={min}
          max={max}
          defaultValue={defaultValue}
          onChange={changeHandler}
        ></input>
        <input
          name={name}
          className="range-input__input"
          ref={inputRef}
          type="number"
          defaultValue={defaultValue}
          min={min}
          max={max}
          onChange={debouncedChangeHandler}
        ></input>
      </div>
      <ClearButton onClick={clearValue} />
    </div>
  );
};
