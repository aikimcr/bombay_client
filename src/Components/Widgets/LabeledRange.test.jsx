import { render } from '@testing-library/react';

import LabeledRange from './LabeledRange.jsx';

jest.useFakeTimers();

it('should show the labeled range', async () => {
  const { asFragment } = render(
    <LabeledRange
      modelName="xyzzy"
      fieldName="plover"
      labelText="plugh"
      min={10}
      max={20}
    />,
  );

  expect(asFragment).toMatchSnapshot();
});

it('should update the number when the slider moves', async () => {
  const result = render(
    <LabeledRange
      modelName="xyzzy"
      fieldName="plover"
      labelText="plugh"
      min={10}
      max={20}
    />,
  );

  const component = result.container.firstChild;
  const sliderField = component.querySelector('input[type="range"]');
  const numberField = component.querySelector('input[type="number"]');

  expect(sliderField).toHaveValue('15');
  expect(numberField).toHaveValue(null);

  await changeInput(component, 'input[type="range"]', 16, 250);
  expect(sliderField).toHaveValue('16');
  expect(numberField).toHaveValue(16);
});

it('should update the slider when the number is changed', async () => {
  const result = render(
    <LabeledRange
      modelName="xyzzy"
      fieldName="plover"
      labelText="plugh"
      min={10}
      max={20}
    />,
  );

  const component = result.container.firstChild;
  const sliderField = component.querySelector('input[type="range"]');
  const numberField = component.querySelector('input[type="number"]');

  expect(sliderField).toHaveValue('15');
  expect(numberField).toHaveValue(null);

  await changeInput(component, 'input[type="number"]', 16, 250);
  expect(sliderField).toHaveValue('16');
  expect(numberField).toHaveValue(16);
});

it('should clear the values', async () => {
  const result = render(
    <LabeledRange
      modelName="xyzzy"
      fieldName="plover"
      labelText="plugh"
      min={10}
      max={20}
    />,
  );

  const component = result.container.firstChild;
  const sliderField = component.querySelector('input[type="range"]');
  const numberField = component.querySelector('input[type="number"]');

  expect(sliderField).toHaveValue('15');
  expect(numberField).toHaveValue(null);

  await changeInput(component, 'input[type="number"]', 11, 250);
  expect(sliderField).toHaveValue('11');
  expect(numberField).toHaveValue(11);

  const clearButton = component.querySelector('button');
  clearButton.click();

  expect(sliderField).toHaveValue('15');
  expect(numberField).toHaveValue(null);
});
