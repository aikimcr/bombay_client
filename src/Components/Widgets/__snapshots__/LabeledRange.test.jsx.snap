// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show the labeled range 1`] = `[Function]`;
